### PR TITLE
feat: input sanitization for event creation & editing

### LINKED ISSUE #450: Input Sanitization for Event Creation & Editing
**Description:**  
Implement input sanitization and validation for all user-facing fields in the event flow (create and edit). This includes title, description, date/time, location, price, capacity, and tags where applicable. The goal is to prevent invalid, malicious, or overly long inputs from being stored or processed.

**Acceptance Criteria:**
- [ ] All event-related inputs are sanitized before being stored or sent to the backend.
- [ ] Attempts to enter HTML, scripts, or SQL-like payloads in event fields are rejected with appropriate error messages.
- [ ] Overly long titles/descriptions/prices/capacities are truncated or rejected according to defined limits.
- [ ] Existing event edit flows cannot bypass required-field validation by clearing mandatory fields.

### PR DESCRIPTION
**Pull request by walid-as**
# Description

## Purpose of the change
This PR implements **comprehensive input sanitization and validation** for all user-facing fields in the event creation and editing flow. The solution prevents **injection attacks**, **malicious inputs**, and enforces **reasonable length limits** while maintaining user experience with **real-time feedback.**

**Key changes include:**
- New `InputSanitizer` utility that protects against XSS, SQL injection, and other dangerous patterns across title, description, price, and capacity fields
- Enhanced field length limits with character counters showing real-time feedback (50 chars for titles, 200 for descriptions)
- Input validation enhancements including prevention of past event dates and improved error messaging
- UI improvements displaying character counts and turning red when limits are reached
- Comprehensive testing covering edge cases, attack patterns, and sanitization scenarios

The implementation follows defense-in-depth principles by sanitizing at input, validating before processing, and providing clear user feedback. All changes are backward compatible and maintain existing form state management patterns.

## Related issues
Closes #450 and part of #452

## How to test
* Check out the branch.  
* Run the app and navigate to event creation or editing screens
* Verify that:
  - Character counters appear for title and description fields.
  - Dangerous input patterns (HTML tags, SQL commands) are blocked.
  - Event dates cannot be set in the past.
  - Price and capacity fields only accept valid numeric input.
  - Existing functionality continues to work normally.
* Run the tests suites to confirm: `InputSanitizerTest.kt` and extended `EditEventFormViewModelTest.kt`.

**Note:** The sanitization is applied in real-time as users type, but validation errors only appear on form submission to avoid disrupting the user experience.
**Note:** ⚠️ This PR slightly exceeds the 500 LOC limit due to the addition of comprehensive security tests for the new input sanitization logic. These tests **(553 LOC)** are critical to ensure robust protection against injection attacks (XSS, SQL injection) and validate that the sanitizer correctly handles edge cases, malicious patterns, and proper error messaging.
**Note:** An LLM was used for the sole and exclusive purpose of improving the clarity and formatting of the PR description.

### CODE REVIEW

File: app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormViewModel.kt [hunk 1]
```diff
@@ -36,6 +37,10 @@ abstract class EventFormViewModel(
 ) : ViewModel() {
   companion object {
     const val MAX_TAG_COUNT = 5
+    const val MAX_TITLE_LENGTH = 50
+    const val MAX_DESCRIPTION_LENGTH = 200
+    const val MAX_PRICE_LENGTH = 10
+    const val MAX_CAPACITY_LENGTH = 6
```
Comment (M-e-d-A) [inline app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormViewModel.kt]: **Question:** Why are `MAX_PRICE_LENGTH = 10` and `MAX_CAPACITY_LENGTH = 6` defined but not enforced in the UI with character counters like title and description? These limits are applied in the sanitization logic. Should we add visual feedback for consistency?

File: app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormViewModel.kt [hunk 2]
```diff
@@ -45,14 +50,18 @@ abstract class EventFormViewModel(
     START_TIME("startTime", "Start time cannot be empty"),
     END_TIME("endTime", "End time cannot be empty"),
     TIME("time", "End time must be after start time"),
+    DATE_IN_PAST("date", "Event cannot be created in the past"),
```
Comment (M-e-d-A) [inline app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormViewModel.kt]: The `DATE_IN_PAST` error message says "Event cannot be created in the past" but this validation applies to both creation and editing. Consider rewording to "Event date cannot be in the past" for accuracy.

File: app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormViewModel.kt [hunk 3]
```diff
@@ -45,14 +50,18 @@ abstract class EventFormViewModel(
...
     PRICE_EMPTY("price", "Price cannot be empty"),
     PRICE_INVALID("price", "Invalid price format"),
     PRICE_NEGATIVE("price", "Price must be positive"),
     CAPACITY_EMPTY("capacity", "Capacity cannot be empty"),
     CAPACITY_INVALID("capacity", "Invalid capacity format"),
-    CAPACITY_NEGATIVE("capacity", "Capacity must be positive");
+    CAPACITY_NEGATIVE("capacity", "Capacity must be positive"),
+    TITLE_DANGEROUS("title", "Title contains invalid characters or dangerous patterns"),
+    DESCRIPTION_DANGEROUS(
+        "description", "Description contains invalid characters or dangerous patterns");
```
Comment (M-e-d-A) [inline app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormViewModel.kt]: **Question:** These error messages say "contains invalid characters or dangerous patterns" which is quite vague for users. Should we provide a hint about what is allowed?

File: app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt [hunk 4]
```diff
@@ -0,0 +1,129 @@
+package ch.onepass.onepass.utils
+
+/** Utility class for sanitizing user input against injection attacks and malicious patterns. */
+object InputSanitizer {
+
+  // Dangerous patterns to detect and reject
+  private val HTML_SCRIPT_PATTERN =
+      Regex(
+          "<script|<img|<iframe|javascript:|onerror=|onclick=|onload=|on\\w+=",
+          RegexOption.IGNORE_CASE)
```
Comment (M-e-d-A) [inline app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt]: The `HTML_SCRIPT_PATTERN` includes `on\w+=` which will match legitimate text like "donation=important". This could cause false positives. Consider making the pattern more specific:

```kotlin
private val HTML_SCRIPT_PATTERN = Regex(
    "<script|<img|<iframe|javascript:|onerror\\s*=|onclick\\s*=|onload\\s*=|on\\w+\\s*=",
    RegexOption.IGNORE_CASE
)
```

File: app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  private val SQL_INJECTION_PATTERN =
+      Regex(
+          "(;\\s*(DROP|DELETE|INSERT|UPDATE|CREATE|ALTER|UNION|SELECT)|--\\s*$|/\\*|\\*/)",
+          RegexOption.IGNORE_CASE)
```
Comment (M-e-d-A) [inline app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt]: The SQL injection pattern catches things like "DROP" and "DELETE" in normal text. For example, a description like "We will DELETE the old process and CREATE a new one" would be rejected. Is this acceptable for your use case? Consider whether you need word boundaries:

```kotlin
private val SQL_INJECTION_PATTERN = Regex(
    "(;\\s*\\b(DROP|DELETE|INSERT|UPDATE|CREATE|ALTER|UNION|SELECT)\\b|--\\s*$|/\\*|\\*/)",
    RegexOption.IGNORE_CASE
)
```

File: app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        .replace(
+            Regex("[\\x00-\\x08\\x0B-\\x0C\\x0E-\\x1F\\x7F]"),
+            "") // Control chars except \n (0x0A) and \t (0x09)
+        .replace("\u200B", "") // Zero-width space
+        .replace("\u200C", "") // Zero-width non-joiner
+        .replace("\u200D", "") // Zero-width joiner
+        .replace("\u202E", "") // Right-to-left override
+        .replace("\u202D", "") // Left-to-right override
+        .replace("\u00AD", "") // Soft hyphen
+  }
```
Comment (M-e-d-A) [inline app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt]: This is very thorough! Good catch on the directional override characters.

File: app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+          parts[0]
+        }
+
+    // Validate decimal places (max 2)
+    if (sanitized.contains(".")) {
+      val decimalParts = sanitized.split(".")
+      if (decimalParts[1].length > 2) {
+        sanitized = "${decimalParts[0]}.${decimalParts[1].take(2)}"
+      }
+    }
```
Comment (M-e-d-A) [inline app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt]: The logic splits on `.` and then rejoins the decimal parts, which means `"12.34.56"` becomes `"12.3456"` before truncating to 2 decimals = `"12.34"`. However, `"12.999.999"` would become `"12.999999"` → `"12.99"`. This behavior is correct but might be unexpected.

File: app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  /**
+   * Sanitizes price input:
+   * - Only digits and decimal point
+   * - No +/- signs
+   * - No scientific notation (e, E)
+   * - Max 2 decimal places
+   */
+  fun sanitizePrice(input: String): String {
```
Comment (M-e-d-A) [inline app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt]: **Question:** If the input has a decimal point but no digits after it (e.g., "12."), the current code will produce "12." as output. Is this intentional? Should it be normalized to "12" instead?

```kotlin
if (decimalParts[1].isEmpty()) {
    sanitized = decimalParts[0]
}
```

**Review by M-e-d-A — CHANGES_REQUESTED** [review]
Thank you for this comprehensive security enhancement PR! The input sanitization implementation is well-structured and addresses critical security concerns. Please review the inline comments below for specific feedback on the code changes.

File: app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormFields.kt
[the remark this answers is not in the document]
Comment (walid-as) [reply app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormFields.kt]: Good one, I extracted the component as requested!

File: app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormViewModel.kt [hunk 9]
```diff
@@ -245,6 +267,13 @@ abstract class EventFormViewModel(
         state.endTime <= state.startTime)
         errors += ValidationError.TIME.toError()
 
+    if (state.date.isNotBlank() && state.startTime.isNotBlank()) {
+      val startTimestamp = parseDateAndTime(state.date, state.startTime)
+      if ((startTimestamp != null) && startTimestamp.toDate().before(java.util.Date())) {
+        errors += ValidationError.DATE_IN_PAST.toError()
+      }
+    }
```
[the remark this answers is not in the document]
Comment (walid-as) [reply app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormViewModel.kt]: That's a great readability improvement! Done.

File: app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormViewModel.kt (hunk 3 above)
Comment (walid-as) [reply app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormViewModel.kt]: I clarified the error message for better UX.

File: app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormViewModel.kt [hunk 10]
```diff
@@ -117,22 +126,33 @@ abstract class EventFormViewModel(
...
   fun updateTitle(title: String) {
-    _formState.value = _formState.value.copy(title = title)
-    clearFieldError(ValidationError.TITLE.key)
+    try {
+      val sanitized = InputSanitizer.sanitizeTitle(title).take(MAX_TITLE_LENGTH)
+      _formState.value = _formState.value.copy(title = sanitized)
+      clearFieldError(ValidationError.TITLE.key, ValidationError.TITLE_DANGEROUS.key)
+    } catch (_: IllegalArgumentException) {
+      _fieldErrors.value = mapOf(ValidationError.TITLE_DANGEROUS.toError())
+    }
```
[the remark this answers is not in the document]
Comment (walid-as) [reply app/src/main/java/ch/onepass/onepass/ui/eventform/EventFormViewModel.kt]: Fixed.

File: app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt (hunk 6 above)
Comment (walid-as) [reply app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt]: Thanks!

File: app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+   * - Removes dangerous patterns
+   * - Limits consecutive newlines to max 2
+   */
+  fun sanitizeDescription(input: String): String {
+    var sanitized = input
+
+    sanitized = removeControlCharacters(sanitized)
+    sanitized = sanitized.replace(Regex(" {2,}"), " ") // Multiple spaces -> single
+    sanitized = sanitized.replace(Regex("\t+"), " ") // Tabs -> space
+    sanitized = sanitized.replace(Regex("\n\n\n+"), "\n\n") // Max 2 newlines
```
[the remark this answers is not in the document]
Comment (walid-as) [reply app/src/main/java/ch/onepass/onepass/utils/InputSanitizer.kt]: Fixed!
